### ISSUE TITLE
Flipper Rate to Motor Speed Conversion Needs a Rescaling Factor

### ISSUE THREAD

**Issue by ESchindewolf** [RoverRobotics/librover#14]
https://github.com/RoverRobotics/librover/blob/214183f083c7363ed706efbde5a47c8db7a9ef23/src/protocol_pro.cpp#L61

The flipper control rate needs to be mapped into the motor speed range in order to impact the value and cause the flipper to move.

Thanks,
-Evan

**Comment by drhieu** [comment]
Could you explain what's not working with this ? Are you using our roverrobotics_ros1 stack to drive the flipper?
It should've been bound to the Right Joystick Vertical Axis. And the Scaling is configurable using the included config file.

**Comment by ESchindewolf** [comment]
Yes, I am using the roverrobotics_ros1 stack for driving the flipper, specifically pro.launch. Also, yes it is tied to the vertical axis of the right joystick. That all traced back fine. The issue I was seeing was that the flipper was acting consistently, but independently of joystick input. I traced the values back through and saw that the drive motor speeds have a *50 for scaling the value into the 0-255 range rather than 0-1 as from the joystick (see below). That's how it appeared to me anyways. I noticed that there was no likewise scaling on the flipper on lines 61 and 62 as mentioned. A scaling factor to map into 0-255 was present in roverrobotics_stack/.../openrover_driver.ccp ([here](https://github.com/RoverRobotics/roverrobotics_stack/blob/b7538bae413b1c4f0332006f3e6612ea6657bebe/rr_openrover_driver/src/openrover_driver.cpp#L444)) so to have that again was my expectation. I used to value of 100 for remapping as in openrover_driver and the flipper controlled as expected. If your intent is to use the config file to cover this mapping, I can multiply the existing value there and should get the same result, correct?

line 132-8 in protocol_pro.cpp
`    motors_speeds_[LEFT_MOTOR] = motor1_control_.boundMotorSpeed(
        int(round(motors_speeds_[LEFT_MOTOR] * 50 + MOTOR_NEUTRAL_)),
        MOTOR_MAX_, MOTOR_MIN_);`
`    motors_speeds_[RIGHT_MOTOR] = motor2_control_.boundMotorSpeed(
        int(round(motors_speeds_[RIGHT_MOTOR] * 50 + MOTOR_NEUTRAL_)),
        MOTOR_MAX_, MOTOR_MIN_);`

Thank you,
-Evan

**Comment by drhieu** [comment]
Hmm that is weird; There should be a third one that would do the same thing as openrover_driver.cpp [here](https://github.com/RoverRobotics/librover/blob/214183f083c7363ed706efbde5a47c8db7a9ef23/src/protocol_pro.cpp#L61-L62)

**Comment by drhieu** [comment]
If you are talking about motor_speed_flipper_coef_ from openrover_driver.cpp , yes we migrated this to be configurable using ros config so you wouldn't have to recompile the code with new value to test with. Otherwise every other variable are defined in the corresponded header.

**Comment by ESchindewolf** [comment]
I moved the remapping factor to the config file as you mentioned and it works fine. I guess I am not understanding the point of having the remapping and scaling factor combined in the config file for the flipper, but only the scaling factor in the config file for the drive motors. If you have a chance to help me understand, that would be great. Otherwise, I think the issue is ready to be closed.

Thanks,
-Evan

**Comment by drhieu** [comment]
Is this [config file](https://github.com/RoverRobotics/roverrobotics_ros1/blob/release/roverrobotics_input_manager/config/controller_config.yaml) not working for you ?. What do you expect to see that is not there? is it the motor_speed_flipper_coef_  that was mentioned? I am a bit confused on where you seeing a **remapping factor** and **scaling factor** on the drive motors

**Comment by ESchindewolf** [comment]
Yes, that file is the one I found and is working now that I've substituted the 0.5 value for 50.0 (0.5 * 100). The way I read this is effectively the scale (0.5) times the remapping (100). Thus essentially incorporating motor_speed_flipper_coef_ into the new system. Sorry for causing confusion. My point before is that with the *50 in lines 132-8 in protocol_pro.cpp the remapping is taken care of for the drive motors in the code and then scales/linear/x in the config file is the scale. At any rate, it works now.
